## CI status: red for a pre-existing base defect — **not** the test flake, and not this PR's code

The `quality` job fails in ~35s **before running a single test**:

```
##[error]An error occurred trying to start process '/usr/bin/bash' with working directory
'/home/runner/work/akgentic-team/akgentic-team/packages/akgentic-team'. No such file or directory
```

This is a **second, distinct** pre-existing defect on the `version-1.3.1` line — separate from the `test_manager.py` on-stop teardown flake described in the PR body. It is an infrastructural checkout failure, not a test failure.

### Evidence that it predates this PR

| Run | Commit under test | Result | Error |
|---|---|---|---|
| [28490832185](https://github.com/b12consulting/akgentic-team/actions/runs/28490832185) | push to bare **`version-1.3.1`** (2026-07-01, no epic-24 code) | fail 31s | byte-identical |
| [29504928781](https://github.com/b12consulting/akgentic-team/actions/runs/29504928781) | `8e39b13` — Amelia's rebased commit, **before** any review fix | fail 38s | byte-identical |
| [29505639357](https://github.com/b12consulting/akgentic-team/actions/runs/29505639357) | `f4465ff` — with review fixes | fail 39s | byte-identical |
| [29503326042](https://github.com/b12consulting/akgentic-team/actions/runs/29503326042) | same story code on the **`master`** base, pre-rebase | **success 1m21s** | — |

The bare base fails identically two weeks before this branch existed, and the same story code passed CI on the `master` base. The variable is the base branch, not the code.

### Root cause

`.github/workflows/ci.yml` on this line checks out the `akgentic-quick-start` workspace pinned to `feat/enterprise-alpha-release-4` (`submodules: recursive`, `secrets.GH_PAT`), then runs the next step with `working-directory: packages/akgentic-team`. That directory never materializes — the recursive submodule checkout is not producing it, which points at the `GH_PAT` secret or the pinned workspace ref rather than at anything in this diff.

Notably, the commit that introduced the pin — *"ci(akgentic-team): pin workspace checkout to feat/enterprise-alpha-release-4"* ([28436275473](https://github.com/b12consulting/akgentic-team/actions/runs/28436275473)) — **also failed**, so this line has been red since the pin landed.

### Why it is not fixed here

Deliberately out of scope, for three reasons:

1. **It is a base defect, not a story defect** — proven byte-identical on the bare base.
2. **No new work goes onto the 1.3.1 line** — the standing decision is that this line carries only the epic's code. A CI-harness change is exactly the kind of new integration that decision excludes.
3. **The likely fix is not in this repo** — it involves a repo secret or the pinned `akgentic-quick-start` workspace ref, so it would cross a submodule boundary.

A re-run was not attempted: the failure is **deterministic**, not flaky — three identical failures across two weeks, including on the bare base.

### Verification actually performed

Since CI cannot exercise this code on this line, the story was verified locally against the rebased base:

- mypy strict — clean, 19 source files
- ruff check (`src/` + `tests/`) — clean
- `EventStore` contract suite — 80 passed / 3 skipped across yaml, mongo, postgres
- full suite — **351–352 passed**, coverage **95.25%** (gate 80%)

The only test failures observed locally were in the known `test_manager.py` on-stop teardown family, which are nondeterministic (a different subset fails on each run) and pass in isolation. See the PR body for that evidence.

**Reported honestly as `ci=flaky-base` rather than claimed green.** This PR should not be treated as CI-verified on the 1.3.1 line until the harness defect is addressed separately.

--- a/src/akgentic/team/repositories/mongo.py
+++ b/src/akgentic/team/repositories/mongo.py
@@ -97,6 +97,12 @@ class MongoEventStore:
         # same database returns silently when an index with the same key spec
         # already exists, so this is safe across redeploys.
         self._teams.create_index("user_id", name="teams_user_id_idx")
+        # ADR-21 §5: backs the load_events(after_event_id=...) anchor lookup.
+        # Not unique — a unique index would turn a read-path ambiguity into a
+        # write-path DuplicateKeyError on save_event.
+        self._events.create_index(
+            [("team_id", 1), ("event.id", 1)], name="events_team_event_id_idx"
+        )
         logger.debug("Initialized MongoEventStore with database '%s'", db.name)
 
     def save_team(self, process: Process) -> None:
@@ -207,6 +213,10 @@ class MongoEventStore:
     ) -> list[PersistedEvent]:
         """Load persisted events for a team, ordered by sequence.
 
+        Both the anchor resolve and the ``sequence > N`` range filter run in
+        MongoDB — the anchor via an indexed, projected ``find_one``, the range
+        as a ``$gt`` clause on the find (ADR-21 §5).
+
         Args:
             team_id: Unique identifier of the team.
             after_event_id: If provided, return only events after the matching
@@ -219,7 +229,10 @@ class MongoEventStore:
             EventNotFoundError: If ``after_event_id`` does not resolve to an
                 event of this team.
         """
-        cursor = self._events.find({"team_id": str(team_id)}).sort("sequence", 1)
+        query: dict[str, object] = {"team_id": str(team_id)}
+        if after_event_id is not None:
+            query["sequence"] = {"$gt": self._resolve_anchor_sequence(team_id, after_event_id)}
+        cursor = self._events.find(query).sort("sequence", 1)
         events: list[PersistedEvent] = []
         for doc in cursor:
             doc.pop("_id", None)
@@ -230,14 +243,27 @@ class MongoEventStore:
                     "Skipping corrupted event for team %s: %s", team_id, exc
                 )
         logger.debug("Loaded %d events for team %s", len(events), team_id)
-        if after_event_id is None:
-            return events
-        # Interim in-memory slice; the $gt push-down lands in story 24-3.
-        # event.id is persisted as a string, so compare stringified ids.
-        for index, event in enumerate(events):
-            if str(event.event.id) == str(after_event_id):
-                return events[index + 1 :]
-        raise EventNotFoundError(f"Event {after_event_id} not found for team {team_id}")
+        return events
+
+    def _resolve_anchor_sequence(self, team_id: uuid.UUID, after_event_id: uuid.UUID) -> int:
+        """Return the ``sequence`` of the cursor anchor, or raise if absent.
+
+        ``event.id`` is persisted as a string, so both ids are coerced with
+        ``str()``: a raw ``uuid.UUID`` would be BSON-encoded as Binary
+        subtype-4 and match nothing. The projection keeps this to a single
+        indexed lookup returning one integer — the document is never hydrated.
+
+        Raises:
+            EventNotFoundError: If the anchor is not an event of this team.
+        """
+        anchor = self._events.find_one(
+            {"team_id": str(team_id), "event.id": str(after_event_id)},
+            projection={"sequence": 1, "_id": 0},
+        )
+        if anchor is None:
+            raise EventNotFoundError(f"Event {after_event_id} not found for team {team_id}")
+        sequence: int = anchor["sequence"]
+        return sequence
 
     def get_max_sequence(self, team_id: uuid.UUID) -> int:
         """Return the highest event sequence number for a team, or 0.

--- a/tests/repositories/mongo/test_mongo.py
+++ b/tests/repositories/mongo/test_mongo.py
@@ -138,6 +138,52 @@ class TestMongoEventStoreMongoSpecific:
         assert info["teams_user_id_idx"]["key"] == [("user_id", 1)]
         assert list(info.keys()).count("teams_user_id_idx") == 1
 
+    # --- event.id index presence --------------------------------------------
+
+    def test_events_team_event_id_index_created_on_init(
+        self, mongo_store: MongoEventStore, mongo_db: Any
+    ) -> None:
+        """``__init__`` creates ``events_team_event_id_idx`` on ``events``.
+
+        Compound ascending key on ``team_id`` + the nested ``event.id``,
+        backing the anchor lookup in ``load_events(after_event_id=...)``
+        (ADR-21 §5). Deliberately not unique: a unique index would turn a
+        read-path ambiguity into a write-path ``DuplicateKeyError`` on
+        ``save_event``.
+        """
+        info = mongo_db["events"].index_information()
+        assert "events_team_event_id_idx" in info
+        assert info["events_team_event_id_idx"]["key"] == [("team_id", 1), ("event.id", 1)]
+        assert not info["events_team_event_id_idx"].get("unique")
+
+    def test_events_team_event_id_index_creation_is_idempotent(
+        self, mongo_store: MongoEventStore, mongo_db: Any
+    ) -> None:
+        """Re-running the constructor against the same database does not raise.
+
+        ``create_index`` returns silently when an index with the same key
+        spec already exists, so redeploys against a long-lived MongoDB are
+        safe. After the second construction the entry is still present
+        exactly once with the same key spec.
+        """
+        # First construction already happened via the ``mongo_store`` fixture.
+        MongoEventStore(mongo_db)  # second construction — must not raise
+
+        info = mongo_db["events"].index_information()
+        assert info["events_team_event_id_idx"]["key"] == [("team_id", 1), ("event.id", 1)]
+        assert list(info.keys()).count("events_team_event_id_idx") == 1
+
+    def test_existing_events_team_sequence_index_still_created(
+        self, mongo_store: MongoEventStore, mongo_db: Any
+    ) -> None:
+        """The pre-existing ``(team_id, sequence)`` index is untouched.
+
+        It already backs the ``{"sequence": {"$gt": n}}`` range filter and
+        needs no change for the cursor push-down.
+        """
+        specs = [entry["key"] for entry in mongo_db["events"].index_information().values()]
+        assert [("team_id", 1), ("sequence", 1)] in specs
+
     # --- Collection name configuration --------------------------------------
 
     def test_collection_names_default_when_env_unset(

--- a/tests/repositories/test_event_store_contract.py
+++ b/tests/repositories/test_event_store_contract.py
@@ -191,6 +191,69 @@ class TestEventStoreContract:
         with pytest.raises(EventNotFoundError):
             event_store.load_events(team_id, after_event_id=uuid.uuid4())
 
+    def test_load_events_first_event_anchor_returns_all_but_first(
+        self, event_store: EventStore
+    ) -> None:
+        """The first event as anchor yields every event except that one."""
+        team_id = uuid.uuid4()
+        for seq in (1, 2, 3, 4, 5):
+            event_store.save_event(make_persisted_event(team_id=team_id, sequence=seq))
+
+        first = event_store.load_events(team_id)[0]
+        tail = event_store.load_events(team_id, after_event_id=first.event.id)
+        assert [e.sequence for e in tail] == [2, 3, 4, 5]
+
+    def test_load_events_last_event_anchor_returns_empty(self, event_store: EventStore) -> None:
+        """The last event as anchor yields ``[]`` — the caller is up to date."""
+        team_id = uuid.uuid4()
+        for seq in (1, 2, 3):
+            event_store.save_event(make_persisted_event(team_id=team_id, sequence=seq))
+
+        last = event_store.load_events(team_id)[-1]
+        assert event_store.load_events(team_id, after_event_id=last.event.id) == []
+
+    def test_load_events_anchor_from_other_team_raises(self, event_store: EventStore) -> None:
+        """An anchor belonging to another team raises — the lookup is team-scoped.
+
+        Returning team A's whole log because the caller passed team B's
+        cursor is the silent degradation ADR-21 §2 exists to prevent.
+        """
+        team_a = uuid.uuid4()
+        team_b = uuid.uuid4()
+        for seq in (1, 2, 3):
+            event_store.save_event(make_persisted_event(team_id=team_a, sequence=seq))
+            event_store.save_event(make_persisted_event(team_id=team_b, sequence=seq))
+
+        foreign_anchor = event_store.load_events(team_b)[0]
+        with pytest.raises(EventNotFoundError):
+            event_store.load_events(team_a, after_event_id=foreign_anchor.event.id)
+
+    def test_load_events_after_event_id_on_empty_store_raises(
+        self, event_store: EventStore
+    ) -> None:
+        """An anchor queried against a team with no events raises, never ``[]``."""
+        with pytest.raises(EventNotFoundError):
+            event_store.load_events(uuid.uuid4(), after_event_id=uuid.uuid4())
+
+    def test_load_events_cursor_round_trip(self, event_store: EventStore) -> None:
+        """The incremental-reader usage pattern: poll, append, poll again.
+
+        This is the case that would catch a silent regression back to a
+        full-log return — the second poll must yield exactly the one new
+        event, not the whole history.
+        """
+        team_id = uuid.uuid4()
+        for seq in (1, 2, 3):
+            event_store.save_event(make_persisted_event(team_id=team_id, sequence=seq))
+
+        cursor = event_store.load_events(team_id)[-1].event.id
+        assert event_store.load_events(team_id, after_event_id=cursor) == []
+
+        event_store.save_event(make_persisted_event(team_id=team_id, sequence=4))
+
+        fresh = event_store.load_events(team_id, after_event_id=cursor)
+        assert [e.sequence for e in fresh] == [4]
+
     # --- get_max_sequence -------------------------------------------------
 
     def test_get_max_sequence_returns_zero_when_empty(self, event_store: EventStore) -> None:

